**MAESTRÍA EN INTELIGENCIA ARTIFICIAL APLICADA**

**Curso: TC5053 - Ciencia y analítica de datos**

Tecnológico de Monterrey

Prof Grettel Barceló Alonso

**Semana 3**
Bases, almacenes y manipulación de datos

---

*   NOMBRE: Alejandro Rendón De Jesús
*   MATRÍCULA: A01664589


---

En esta actividad usarás una base de datos relacional basada en el informe de participación y la lista del top 10 de Netflix. Incluye películas y programas de televisión, así como información sobre temporadas, métricas de visualización, fechas de estreno, duración y más, organizada en las siguientes tablas:

* `movie`: Información general de las películas.

* `tv_show`: Información general de los programas de televisión.

* `season`: Datos de las temporadas asociadas a cada programa de TV.

* `view_summary`: Métricas de visualización y rendimiento de películas o temporadas.

Revisa con detalle su esquema para que comprendas cómo se relacionan las tablas anteriores.

**NOTA IMPORTANTE:** Asegúrate de responder *explícitamente* todos los cuestionamientos.


`PyMySQL` es una librería escrita en Python puro que funciona como conector (*driver*) para motores de bases de datos MySQL, permitiendo abrir conexiones, ejecutar consultas SQL y recuperar resultados directamente desde programas en Python.

`SQLAlchemy` es una librería de Python que facilita la interacción con bases de datos y permite mantener un pool de conexiones eficiente, gestionar *commits* y *rollbacks* de forma automática y asegurar que múltiples conexiones simultáneas se manejen de manera segura, incluso cuando se ejecutan consultas SQL “puras”

In [1]:
# Importa las librerías necesarias
import pymysql
import sqlalchemy as sqla
import pandas as pd

Se crea una conexión (`conn`) para luego invocar declaraciones SQL.

In [2]:
# motor+driver://usuarioBD:clave@ipHostDBMS:puerto/esquema
# pool_recycle controla el tiempo máximo de vida de una conexión en el pool (3600 segundos = 1 hora)
db = sqla.create_engine('mysql+pymysql://mnaTC5053User:mnaTC5053Pass!@20.127.169.219:3306/Netflix', pool_recycle=3600)
conn = db.connect()

Para que tus consultas sean más legibles y fáciles de mantener, puedes usar este formato multilínea con comillas triples y `sqla.text()`. Por ejemplo:

```
query = sqla.text("""
  SELECT ---
  FROM ---
  WHERE ---
""")
pd.read_sql(query, conn)
```

In [3]:
def sql(query, params = {}):
    global conn 

    df = pd.read_sql(
        sqla.text(query),
        conn, 
        params = params
    )

    return df

1.	Extrae toda la información de las películas que duran más de 5 horas.

In [4]:
df1 = sql(
    """ 
    SELECT 
        *
    FROM 
        movie 
    WHERE 
        COALESCE(runtime / 60, 0) >= 5 
    ORDER BY 
        runtime DESC 
    """
)
df1

,id,created_date,modified_date,available_globally,locale,original_title,release_date,runtime,title
0,5793,2024-01-01,2024-01-01,b'\x00',None,日本統一シリーズ: 映画シリーズ,None,3892,Nihontouitsu Series: Film Series
1,5794,2024-01-01,2024-01-01,b'\x00',None,釣りバカ日誌: 映画シリーズ,None,2120,Free and Easy Series: Film Series
2,9729,2024-01-01,2024-01-01,b'\x00',None,織田同志会 織田征仁: 映画シリーズ,None,710,Seiji Oda: Film Series
3,9730,2024-01-01,2024-01-01,b'\x00',None,キングダム～首領になった男～: 映画シリーズ,None,427,Kingdom ~ The Man Who Became the Top ~: Film S...
4,5874,2024-01-01,2024-01-01,b'\x01',None,None,2021-08-06,312,Navarasa: Limited Series


2.	¿Cuál es el porcentaje de películas disponibles únicamente en EU en relación con el total, excluyendo los valores `NULL`? La consulta SQL debe entregar directamente el porcentaje final. No se deben devolver resultados parciales para realizar el cálculo en Python.

In [5]:
df2 = sql(
    """
    SELECT
        100 * SUM(available_globally = 0) / COUNT(available_globally) AS only_US_p
    FROM
        movie
    """
)
df2.set_index([['%']])

,only_US_p
%,83.5717


**Respuesta:** el **83.57 %** de las películas está disponible únicamente en EU (`available_globally = 0`), considerando solo los registros donde `available_globally` no es `NULL` (`COUNT(columna)` ignora los `NULL`, por lo que el denominador ya los excluye). El 16.43 % restante corresponde a las películas disponibles globalmente.

3.	¿Cuáles son los idiomas o regiones originales en la tabla de películas?
* ¿Cuántos registros tienen el campo `locale` con valor `NULL`? (NULL en SQL ⇔ None en Python)

In [6]:
df31 = sql(
    """ 
    SELECT DISTINCT locale FROM movie
    WHERE locale IS NOT NULL
    """
)

df31

,locale
0,en


In [7]:
df32 = sql(
    """ 
    SELECT 
        COUNT(CASE WHEN locale IS NULL THEN 1 END) AS N_locale_nulls
    FROM 
        movie
    """
)

df32

,N_locale_nulls
0,11255


**Respuesta:** el único idioma/región original registrado en `movie` es **`en`** (inglés). Hay **11,255** registros con `locale` en `NULL` (de 11,787 películas), es decir, solo 532 tienen el campo capturado.

4.	Asumiendo que los valores `NULL` en `locale` corresponden a otro idioma (diferente del inglés), el título original de la película NO debería coincidir con el título principal en dichos registros.
* Determina cuántas películas tienen títulos diferentes en estos dos campos (`title` y `original_title`).
*  ¿Coinciden los resultados (cantidad de `NULL` y títulos diferentes)? Si no es así, identifica qué características tienen los registros restantes.
* Finalmente, concluye si la suposición de que los valores `NULL` en `locale` indican que la película está en otro idioma es válida.

In [8]:
df41 = sql(
    """
    SELECT 
        COUNT(*) AS N_movies_different_title
    FROM 
        movie
    WHERE
        title != original_title
    """
)

df41

,N_movies_different_title
0,3947


In [9]:
# Registros con locale NULL que NO tienen títulos diferentes: ¿qué características tienen?
df42 = sql(
    """
    SELECT
        CASE
            WHEN original_title IS NULL THEN 'original_title NULL'
            WHEN title = original_title THEN 'title = original_title'
        END AS caracteristica,
        COUNT(*) AS N_movies
    FROM
        movie
    WHERE
        locale IS NULL
        AND (original_title IS NULL OR title = original_title)
    GROUP BY
        caracteristica
    """
)

df42

,caracteristica,N_movies
0,original_title NULL,7291
1,title = original_title,17


In [10]:
# Los registros donde ambos títulos "coinciden"
df43 = sql(
    """
    SELECT
        id, title, original_title
    FROM
        movie
    WHERE
        locale IS NULL
        AND title = original_title
    """
)

df43

,id,title,original_title
0,1089,Recep Ivedik 7,Recep İvedik 7
1,1349,Veronica,Verónica
2,3509,Fantomas,Fantômas
3,3611,Luccas Neto em: O Hotel Magico,Luccas Neto em: O Hotel Mágico
4,4566,Un Dia En El Paraiso,Un día en el paraíso
5,5077,Superlopez,Superlópez
6,5211,Baris Akarsu Merhaba,Barış Akarsu Merhaba
7,5684,Totem,Tótem
8,6427,La Pena Maxima,La pena máxima
9,6530,Mirciulica,Mirciulică


In [11]:
# Ejemplos de registros con locale NULL y sin título original
df44 = sql(
    """
    SELECT
        id, title, original_title, locale
    FROM
        movie
    WHERE
        locale IS NULL
        AND original_title IS NULL
    ORDER BY
        id
    LIMIT 10
    """
)

df44

,id,title,original_title,locale
0,1,Damsel,None,None
1,2,Lift,None,None
2,5,The Super Mario Bros. Movie,None,None
3,10,The Boss Baby,None,None
4,16,Leo,None,None
5,21,Trigger Warning,None,None
6,30,Sing 2,None,None
7,36,Heart of the Hunter,None,None
8,37,Despicable Me 2,None,None
9,38,Sing (2016),None,None


**Respuestas:**

* Hay **3,947** películas con `title` distinto de `original_title`.
* **No coinciden**: hay 11,255 registros con `locale` en `NULL` y solo 3,947 con títulos diferentes (todos ellos con `locale` `NULL`). Los **7,308** registros restantes tienen estas características:
  * **7,291** tienen `original_title` en `NULL`, por lo que la comparación `title != original_title` da `NULL` y no se cuentan. Entre ellos hay películas claramente en inglés (*Damsel*, *Lift*, *The Super Mario Bros. Movie*, *The Boss Baby*…).
  * **17** tienen ambos títulos "iguales" para MySQL, pero en realidad solo difieren en acentos/diacríticos (*Veronica* / *Verónica*, *Tiramisu* / *Tiramisù*); la intercalación de la columna no distingue acentos, así que los trata como iguales. Son películas en otro idioma.
* **Conclusión:** la suposición **no es válida**. Un `locale` `NULL` significa que el dato no fue capturado, no que la película esté en otro idioma: solo ~35 % de esos registros (3,947 + 17 de 11,255) muestra evidencia de un título original distinto, y el otro ~65 % no tiene título original e incluye muchas películas en inglés.

5.	Determina el título de la película que ha permanecido más tiempo en el top 10.

In [12]:
df5 = sql(
    """
    SELECT
        m.title,
        MAX(v.cumulative_weeks_in_top10) AS weeks_in_top10
    FROM
        view_summary v
        JOIN movie m ON m.id = v.movie_id
    GROUP BY
        m.id, m.title
    ORDER BY
        weeks_in_top10 DESC
    LIMIT 1
    """
)

df5

,title,weeks_in_top10
0,The Boss Baby,22


**Respuesta:** la película que más tiempo ha permanecido en el top 10 es ***The Boss Baby***, con **22 semanas** acumuladas (`cumulative_weeks_in_top10`).

6.	Identifica los 5 programas de TV con mayor cantidad de temporadas.

In [13]:
df6 = sql(
    """
    SELECT
        t.title,
        COUNT(s.id) AS N_seasons
    FROM
        tv_show t
        JOIN season s ON s.tv_show_id = t.id
    GROUP BY
        t.id, t.title
    ORDER BY
        N_seasons DESC, t.title
    LIMIT 5
    """
)

df6

,title,N_seasons
0,Grey's Anatomy,20
1,Naruto Shippuden,20
2,Heartland (2007),17
3,It's Always Sunny in Philadelphia,16
4,NCIS,15


**Respuesta:** los 5 programas con más temporadas son *Grey's Anatomy* (20), *Naruto Shippuden* (20), *Heartland (2007)* (17), *It's Always Sunny in Philadelphia* (16) y *NCIS* (15).

Nota: el quinto lugar está empatado con 15 temporadas entre *NCIS*, *Raw* y *Supernatural (2005)*; el empate se resolvió por orden alfabético.

7.	¿Cuáles son los intervalos de fechas de los resúmenes en la tabla `view_summary` de los períodos (`duration`) semestrales?

In [14]:
df7 = sql(
    """
    SELECT DISTINCT
        start_date,
        end_date
    FROM
        view_summary
    WHERE
        duration = 'SEMI_ANNUALLY'
    ORDER BY
        start_date
    """
)

df7

,start_date,end_date
0,2023-07-01,2023-12-31
1,2024-01-01,2024-06-30


**Respuesta:** los resúmenes semestrales cubren dos intervalos:

* del **2023-07-01 al 2023-12-31** (segundo semestre de 2023)
* del **2024-01-01 al 2024-06-30** (primer semestre de 2024)

8.	Ordena las temporadas de *Grey's Anatomy* según la cantidad de vistas registradas en el primer período semestral de 2024.
* ¿Cómo interpretarías los resultados obtenidos?

In [15]:
df8 = sql(
    """
    SELECT
        s.season_number,
        s.title,
        v.views,
        v.hours_viewed
    FROM
        tv_show t
        JOIN season s ON s.tv_show_id = t.id
        JOIN view_summary v ON v.season_id = s.id
    WHERE
        t.title = 'Grey''s Anatomy'
        AND v.duration = 'SEMI_ANNUALLY'
        AND v.start_date = '2024-01-01'
        AND v.end_date = '2024-06-30'
    ORDER BY
        v.views DESC
    """
)

df8

,season_number,title,views,hours_viewed
0,1,Grey's Anatomy: Season 1,3600000,23400000
1,2,Grey's Anatomy: Season 2,3100000,60600000
2,3,Grey's Anatomy: Season 3,2900000,53700000
3,5,Grey's Anatomy: Season 5,2900000,49300000
4,4,Grey's Anatomy: Season 4,2900000,35300000
5,6,Grey's Anatomy: Season 6,2800000,48000000
6,7,Grey's Anatomy: Season 7,2700000,43000000
7,8,Grey's Anatomy: Season 8,2600000,44700000
8,9,Grey's Anatomy: Season 9,2500000,43300000
9,10,Grey's Anatomy: Season 10,2400000,41700000


**Interpretación:**

* Las vistas **decrecen casi de forma monótona con el número de temporada**: la temporada 1 es la más vista (3.6 M) y de ahí bajan gradualmente hasta ~1.7 M en las temporadas 17 y 18. Es el patrón típico de embudo: mucha gente empieza la serie desde el inicio y la va abandonando conforme avanza.
* La **temporada 19 rompe la tendencia** (2.0 M, por encima de las temporadas 15 a 18), lo que sugiere audiencia que sigue la serie al día y ve directamente las temporadas más recientes.
* La **temporada 20 queda muy por debajo** (0.1 M). No tiene registro en el semestre anterior, así que lo más probable es que se haya agregado al catálogo al final del período y apenas acumuló horas; no refleja falta de interés.
* Las vistas no son lo mismo que las horas vistas: la temporada 1 lidera en vistas pero es de las que menos horas acumula (23.4 M), porque es mucho más corta; en horas vistas lidera la temporada 2 (60.6 M).

9.	Todas las consultas anteriores podrían hacerse también con la lógica relacional implementada en Pandas, que permite replicar la mayoría de las operaciones de SQL. Si los dataframes se han llenado como sigue, resuelve la consulta 8 utilizando las funciones de Pandas.

In [16]:
tv_show = pd.read_sql(sqla.text('SELECT * FROM tv_show'), conn)
season = pd.read_sql(sqla.text('SELECT * FROM season'), conn)
view_summary = pd.read_sql(sqla.text('SELECT * FROM view_summary'), conn)

In [17]:
greys_id = tv_show.loc[tv_show['title'] == "Grey's Anatomy", 'id']

df9 = (
    season[season['tv_show_id'].isin(greys_id)]
    .merge(
        view_summary,
        left_on = 'id',
        right_on = 'season_id',
        suffixes = ('_season', '_view')
    )
    .query("duration == 'SEMI_ANNUALLY'")
)

df9 = df9[
    (pd.to_datetime(df9['start_date']) == '2024-01-01')
    & (pd.to_datetime(df9['end_date']) == '2024-06-30')
]

df9 = (
    df9[['season_number', 'title', 'views', 'hours_viewed']]
    .sort_values('views', ascending = False, kind = 'stable')
    .reset_index(drop = True)
    .astype({'season_number': int})
)

df9

,season_number,title,views,hours_viewed
0,1,Grey's Anatomy: Season 1,3600000,23400000
1,2,Grey's Anatomy: Season 2,3100000,60600000
2,3,Grey's Anatomy: Season 3,2900000,53700000
3,5,Grey's Anatomy: Season 5,2900000,49300000
4,4,Grey's Anatomy: Season 4,2900000,35300000
5,6,Grey's Anatomy: Season 6,2800000,48000000
6,7,Grey's Anatomy: Season 7,2700000,43000000
7,8,Grey's Anatomy: Season 8,2600000,44700000
8,9,Grey's Anatomy: Season 9,2500000,43300000
9,10,Grey's Anatomy: Season 10,2400000,41700000


`MySQL` es un sistema de gestión de bases de datos relacional, pero desde Python también es posible extraer información de bases de datos no relacionales, como `Firestore`, `MongoDB` o `Cassandra`, utilizando conectores o integraciones específicas. Esto permite combinar datos de diferentes fuentes según las necesidades del análisis o la aplicación.

En el siguiente ejercicio usarás un ejemplo con `Firestore` desde Python. Para ello utilizarás los módulos `credentials` y `firestore` de la biblioteca `firebase_admin`.

In [18]:
import firebase_admin
from firebase_admin import credentials
from firebase_admin import firestore

En `Firestore`, a diferencia de `MySQL` donde se utiliza un usuario y contraseña para conectarse, la autenticación se realiza mediante un archivo JSON que almacena las credenciales necesarias para acceder a la base de datos. Este archivo contiene las claves y la información de configuración que permiten a Python establecer la conexión de manera segura.

In [19]:
# consultancy.json almacena la clave privada para autenticar una cuenta y autorizar el acceso a los servicios
# A través de la función Certificate(), se regresa una credencial inicializada, que puedes utilizar para crear una nueva instancia de la aplicación
cred = credentials.Certificate('consultancy.json')
firebase_admin.initialize_app(cred)
db = firestore.client()

10.	Investiga cómo leer la colección `EMPLOYEE` y mostrar su contenido en un dataframe. Asegúrate de incluir el `id` en el resultado

In [20]:
# stream() recupera todos los documentos de la colección;
# cada documento expone su identificador en .id y sus campos con .to_dict()
docs = db.collection('EMPLOYEE').stream()

employees = pd.DataFrame([{'id': doc.id, **doc.to_dict()} for doc in docs])
employees

,id,emp_fname,emp_lname,emp_hiredate
0,8LcLuxVHGAd2d9IQc5jF,David,Senior,1989-07-12 06:00:00+00:00
1,Fzd60D6Z2CU4n0wVV8YN,William,Smithfield,2004-06-04 05:00:00+00:00
2,lX5xuQ5w3i6ib2ExccWY,John,News,2000-11-08 06:00:00+00:00
3,yocFj2lichOkbAj9NBfp,June,Arbough,1996-12-01 06:00:00+00:00


In [21]:
firebase_admin.delete_app(firebase_admin.get_app())

---

**Declaración de uso de IA**

Si aplica, deberá indicarse la herramienta y el modelo empleado en la entrega, así como la finalidad de su uso (generación de código / depuración / optimización).

Por ejemplo:

*   OpenAI. (2026). *ChatGPT (basado en GPT-4)* [Modelo de lenguaje grande], utilizado para generación de código y depuración. https://chat.openai.com/

---